torch/distributed/_tensor/experimental/local_map.py

```diff
@@ -139,44 +157,62 @@ def wrapped(*args, **kwargs):
                             arg = arg.redistribute(device_mesh, spec)
                         else:
                             raise ValueError(
-                                f"arg {arg} in local_map has a mismatched placements:"
-                                f"arg placements is {arg.placements} but the input"
-                                f"placements is {spec}!"
-                                "If redistribute_inputs is wanted, set redistribute_inputs=True to local_map."
+                                f"arg {arg} in local_map has a mismatched placements: "
+                                f"arg placements is {arg.placements} but the input "
+                                f"placements is {spec}! "
+                                "If redistribute_inputs is wanted, set "
+                                "redistribute_inputs=True to local_map."
                             )
 
-                flat_local_args.append(arg.to_local())
+                local_arg = arg.to_local()
+                if isinstance(local_arg, AsyncCollectiveTensor):
+                    local_arg = local_arg.wait()
+
+                flat_local_args.append(local_arg)
             else:
+                # Non-Tensor input must have None in `in_placements`
+                if in_placements is not None and not isinstance(arg, torch.Tensor):
+                    spec = in_placements[idx]
+                    assert spec is None, (
+                        f"Non-Tensor input {arg} expects None placements "
+                        f"but received {spec}!"
+                    )
+
                 flat_local_args.append(arg)
 
         local_args = pytree.tree_unflatten(flat_local_args, args_spec)
 
-        out = func(device_mesh, *local_args, **kwargs)
+        out = func(*local_args, **kwargs)
 
-        # process output
-        flat_out, out_spec = pytree.tree_flatten(out)
-        flat_dist_out = []
-        for idx, out in enumerate(flat_out):
-            spec = (
-                out_placements[idx]
-                if isinstance(out_placements, tuple)
-                else out_placements
-            )
-            if isinstance(out, torch.Tensor):
-                assert not isinstance(
-                    out, DTensor
-                ), f"torch.Tensor output expected but received {type(out)}: {out}"
+        if seen_dtensor_arg:
+            # process output
+            flat_out, out_spec = pytree.tree_flatten(out)
 
-                flat_dist_out.append(
-                    DTensor.from_local(out, device_mesh, spec, run_check=False)
+            flat_dist_out = []
+            for idx, out in enumerate(flat_out):
+                spec = (
+                    out_placements[idx]
+                    if isinstance(out_placements, tuple)
```

Comment:
we should have `out_placements = tuple(out_placements) if not isinstance(out_placements, tuple)` first, and then in this for loop we can simply `for out, out_placement in zip(flat_out, out_placements)`?

Reply:
That sounds better, let me address it in a follow up along with adding the test.

Reply:
comment addressed in #127752